codice/ddf
# [2.26.x] Improve audit logs for authentication and session events

#### What does this PR do?
Cleans up DDF's audit logs for authentication and session events:
* Login success/failure
* Logout
* Session creation/deletion/expiration
* Invalid session

#### Who is reviewing it? 

@glenhein @jlcsmith 

#### Select relevant component teams: 
@codice/security 

#### How should this be tested?

Perform each of the following actions and make sure the output matches what is shown. The auth actions (login, logout) should be tested for BASIC, PKI, and OIDC except authentication failure - OIDC will use an external auth server so DDF won't know when the user login fails. Note: the audit logs may differ slightly between auth types, but all the info should be there. 
 
* Log in
```
[INFO ] 2020-12-10T12:05:49,646 | qtp529305888-135 | securityLogger  |  Subject: admin Logged in
[INFO ] 2020-12-10T12:05:49,671 | qtp529305888-135 | securityLogger  |  Subject: admin Creating a new session with id 5826261cb8e381cb765659c57fcf4d426949bbba5b4079e67fa4d6fc51fbcdc8 for client 0:0:0:0:0:0:0:1.
```
* Log out. Make sure the session id matches the one created when you logged in
```
[INFO ] 2020-12-10T12:05:59,218 | qtp529305888-136 | securityLogger  |  Subject: UNKNOWN Subject admin logged out
[INFO ] 2020-12-10T12:05:59,227 | qtp529305888-136 | securityLogger  |  Subject: UNKNOWN Session 5826261cb8e381cb765659c57fcf4d426949bbba5b4079e67fa4d6fc51fbcdc8 destroyed
```
* Try to log in with bad credentials (Basic auth)
```
[INFO ] 2020-12-10T12:15:31,951 | qtp529305888-448 | securityLogger  |  Subject: UNKNOWN Authentication failed. Error message: 'Login failed for user: admin'
```
* In the Admin UI, change the session timeout (Security > Session > Session Timeout) to 2 minutes. Then log in and wait for the session to expire. Jetty won't delete it immediately after 2 minutes. It seems to expire sessions in batches every 10-15 minutes or so. Make sure the session created when you logged in matches the expired session. 
```
[INFO ] 2020-12-10T12:08:50,450 | qtp529305888-135 | securityLogger  |  Subject: admin Logged in
[INFO ] 2020-12-10T12:08:50,489 | qtp529305888-136 | securityLogger  |  Subject: admin Creating a new session with id 5cbd4636fbbcf75580d435080e25b77e9f35426fcb3a98782f78c6c687f40bbc for client 0:0:0:0:0:0:0:1.
...
[INFO ] 2020-12-10T12:24:52,906 | eeper-4fc9a898-1 | securityLogger  |  Subject: UNKNOWN Session 5cbd4636fbbcf75580d435080e25b77e9f35426fcb3a98782f78c6c687f40bbc expired
``` 
* Create a request with an invalid session id `curl -k --cookie "JSESSIONID=123" https://localhost:8993/`
```
[INFO ] 2020-12-10T15:36:40,198 | qtp529305888-507 | securityLogger  |  Subject: UNKNOWN Request contained invalid or expired session id [a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3]
```

#### Any background context you want to provide?

#### What are the relevant tickets?
Fixes: #____

#### Screenshots


#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
